added files for binary builds on openSUSE build service

mbhangui opened the pull request:
These files have been added for openSUSE build service on https://build.opensuse.org/project/show/home:notqmail

Description of the important files
notqmail.spec - The RPM spec file
notqmail-rpmlintrc - Containing rules to ignore few non-compliance errors (mostly FHS) on obs.
notqmail-permission.* - These files are needed for SUSE Enterprise Linux (AFAIK)
qmail-send.service - systemd unit file to start qmail-send
notqmail.dsc - packages that opensuse build service should pull before doing a notqmail build
debian/rules - The debian rules file
deban/control - debian packaging control file
debian/preinst,postinst, prerm, postrm - debian pre/post installation and deinstallation scripts

alanpost commented:
It's certainly fine to use the same PR, but will you separate the debian and redhat packaging in to separate commits?

alanpost commented:
The debian/ directory belongs in the root of notqmail.git, not inside the open the opensuse_buildservice directory.  That is the location it is expected when you run dpkg-buildpackage and friends locally.

mbhangui commented:
> The debian/ directory belongs in the root of notqmail.git, not inside the open the opensuse_buildservice directory. That is the location it is expected when you run dpkg-buildpackage and friends locally.

OK. I will move the the spec file and the debian subdirectory to the root.

There are two ways of doing it. One is having debian subdirectory in the netqmail source tar ball and the other is by having a single tar file named debian.tar.gz. If I have the debian directory in the root, one cannot have debian.tar.gz. My plan is to do the following

1. Modify Makefile to create debian.tar.gz in the opensuse_buildservice directory
2. Have a _service file on opensuse build service notqmail directory which will pull notqmail.spec, notqmail.dsc, notqmail-rpmlintr, notqmail-permssion.* files, debian.dsc and debian.tar.gz. obs will then automatically pull all these files when triggerred using the web ui or using the osc command.
3. A way to automatically trigger the build when anything is updated (just like travis). I am yet to figure this out.
4. A script to automatically create the changelog from the CHANGES file.

Let me think of a way where both can be achieved and I will reorganize the files accordingly

1. Building the RPM or debian locally so that any user can build the packages locally by just executing the commands rpmbuild or dpkg-buildpackage
and
2. Building the RPM and debian packages on the obs

schmonz commented:
Bump, @mbhangui and others. How can we help get this done and merged for 1.08?

mbhangui commented:
> Bump, @mbhangui and others. How can we help get this done and merged for 1.08?
I just need to upload the latest tar.gz and do the build. I will do this tomorrow

DerDakon commented:
Maybe just wait until 1.08 is out, the build is done and we are satisfied with it, and then merge it. I would like to avoid that we have a bunch of 1.07.9999 build with patches or weird tarballs and all that. YMMV.

DerDakon commented:
I took the opportunity to rebase this on current master, adapt to some 1.08 changes, and add a second optional commit that drops some things that we want to remove in 1.09 even from the 1.08 packages.

DerDakon commented:
I upgraded one of my servers using this RPM this morning and everything went fine. Some thoughts:

-is it the default to enable a newly installed service?
-what happens if a user like me overrides the service by placing a modified copy in /etc/systemd, should we disable this on package removal?
-/var/qmail/queue/lock/tcpto will show as modified when using "rpm --verify notqmail" because it is modified by the running services. Is there a way that we can mark it as "this file will be modified"? Mark as configuration file?

mbhangui commented:
> -is it the default to enable a newly installed service?
Yes it is the default with almost all standard daemons (atleast on Fedora & CentOS). I have once in a while tried out Ubuntu and Debian. There too, these services get installed for startup by default. A good example is MySQL.

> -what happens if a user like me overrides the service by placing a modified copy in /etc/systemd, should we disable this on package removal?

I don't have a definite answer to this. But in postinst, most packages do
```
systemctl disable servicename
```
> -/var/qmail/queue/lock/tcpto will show as modified when using "rpm --verify notqmail" because it is modified by the running services. Is there a way that we can mark it as "this file will be modified"? Mark as configuration file?

I had this issue in indimail-mta. So what I did was

1. have /var/indimail/queue in the spec file
2. create the queue during postinstall of the package
3. Files which can be created later but actuall belong to the package can be marked as ghost. e.g.
```
%ghost %config(noreplace,missingok)                %{qsysconfdir}/tcp/tcp.smtp
```
Today I have built successfully binary builds for all the distros configured in OBS. Added Ubuntu 20.04 and openSUSE leap 15.2. I have done some modifications to the spec file and tested the installation on ubuntu 20.04 and CentOS7. I also tested the delivery. I had to make changes to fix few errors. The errors were all related to installing the systemd unit file.

Meanwhile I have also created docker & podman container images for notqmail. I will create a pull request for the same. The podman/docker images will use the opensuse build service to build the container.

mbhangui commented:
Oh I forgot to mention. I have created a _service file for obs. Having such a file enables obs to pull the source archive directly from github
```
<services>
  <service name="download_url">
    <param
        name="host">github.com</param>
    <param name="protocol">https</param>
    <param name="path">/notqmail/notqmail/releases/download/notqmail-1.08/notqmail-1.08.tar.xz</param>
  </service>
</services>
```

mbhangui commented:
I did some screw up. I pushed one commit which was supposed to be committed in the opensuse_buildservice branch to the docker branch. I don't know how to fix it

DerDakon commented:
You can do something like "git checkout docker; git reset --hard HEAD^; git push --force" which would wipe the last commit from that given branch.

mbhangui commented:
Yes. That worked

mbhangui commented:
@DerDakon  can you check the docker pull request. I see older commits that were related to opensuse_buildservice there. I don't understand how those have gone there because the docker pull request is something I just created few minutes back.

I have created the dockerfiles created for ubuntu and centos but don't want to push it till this is cleaned up.

DerDakon commented on opensuse_buildservice/debian/copyright:
```diff
@@ -0,0 +1,41 @@
+netqmail 1.06 to notqmail 1.07.
+NOTE: notqmail 1.07 is a community-driven fork of qmail.
+Neither D. J. Bernstein nor any of the netqmail developers has
+participated in or been asked to approve of this distribution.
```
1.08 is missing here.

mbhangui replied:
Yes will fix this

DerDakon commented on opensuse_buildservice/debian/postinst:
```diff
@@ -0,0 +1,102 @@
+
+qmaildir=/var/qmail
+logdir=/var/log/svc
+cp=/bin/cp
```
Are these really needed or can't we use just "cp" and expect it to be in the path?

mbhangui replied:
I guess this can be removed. I have this habit of using the system provided cp in shell scripts, to avoid a trojan cp getting executed. I had admins in my company using $HOME/bin, ., etc in the PATH first.

DerDakon replied:
since rpm is run as root this shouldn't be an issue.

DerDakon commented on opensuse_buildservice/debian/postinst:
```diff
@@ -0,0 +1,102 @@
+
+case "$1" in
+    configure)
+		if [ " $2" = " " ] ; then
```
The indentation here is somewhat broken.

DerDakon commented on opensuse_buildservice/debian/postinst:
```diff
@@ -0,0 +1,102 @@
+echo "Post Installation..."
+(
+echo "Creating ${logdir}"
+if [ ! -d ${logdir} ] ; then
```
Is this check needed? ```mkdir -p``` will just do nothing if the directory already exists.

mbhangui replied:
Yes

DerDakon commented on opensuse_buildservice/debian/postrm:
```diff
@@ -0,0 +1,102 @@
+		${rmdir} --ignore-fail-on-non-empty ${qmaildir}/alias 2>/dev/null || true
+	fi
+	if [ -h ${logdir} ] ; then
+		log_dir=`/bin/ls -ld ${logdir} | /usr/bin/awk '{print $10}'`
```
readlink?

mbhangui replied:
Yes. Will modify this to use readlink

DerDakon commented on opensuse_buildservice/debian/preinst:
```diff
@@ -0,0 +1,70 @@
+		${rm} -f /var/spool/mail/$i
+	done
+	
+	/usr/bin/getent passwd alias    > /dev/null || /usr/sbin/useradd -M -g nofiles  -d ${qmaildir}/alias  -s /sbin/nologin  alias || true
```
Seems like Debain policies prefer ```adduser```

mbhangui replied:
I'm not sure and knowledgeable about this. I will check this out.

mbhangui replied:
done for addgroup and adduser

DerDakon commented on opensuse_buildservice/notqmail.spec:
```diff
@@ -0,0 +1,423 @@
+/usr/bin/getent passwd qmailq   > /dev/null || /usr/sbin/useradd -M -g qmail    -d %{qmaildir}        -s /sbin/nologin qmailq
+/usr/bin/getent passwd qmailr   > /dev/null || /usr/sbin/useradd -M -g qmail    -d %{qmaildir}        -s /sbin/nologin qmailr
+/usr/bin/getent passwd qmails   > /dev/null || /usr/sbin/useradd -M -g qmail    -d %{qmaildir}        -s /sbin/nologin qmails
+for i in alias qmaild qmaill qmailp qmailq qmailr qmails qscand
```
qscand?

mbhangui replied:
Oops. Result of copy paste. This has to go away

DerDakon commented on opensuse_buildservice/debian/rules:
```diff
@@ -0,0 +1,100 @@
+clean:
+	rm -f build-stamp configure-stamp
+	${rm} -rf $(DESTDIR)
+	$(MAKE) -s clean||true
```
Why can this fail?

mbhangui replied:
It can be just $(MAKE) is clean. This check is there because of copy-paste from something that uses autoconf, automake, etc, where make distclean removes the Makefile and hence causes make command to fail.

mbhangui replied:
I will remove || true as notmail's Makefile doesn't have distclean

mbhangui replied:
This got left out. But I see no harm. Else I will have to do another update

DerDakon commented on opensuse_buildservice/qmail-send.service:
```diff
@@ -0,0 +1,11 @@
+[Unit]
```
This can now go away, the repository already has such a file. ```make qmail-send.service``` will create it.

mbhangui replied:
OK

DerDakon commented on opensuse_buildservice/debian/rules:
```diff
@@ -0,0 +1,100 @@
+	env DESTDIR=$(DESTDIR) ./instpackage
+	${cp} COPYRIGHT $(DESTDIR)$(qmaildir)/doc
+	${cp} instchown $(DESTDIR)$(qmaildir)/bin
+	( \
```
Use ```make qmail-send.service``` to create the service file, then install it whereever it should go.

mbhangui replied:
OK

DerDakon commented on opensuse_buildservice/notqmail-rpmlintrc:
```diff
@@ -0,0 +1,21 @@
+addFilter("W: suse-filelist-forbidden-fhs23 /var/qmail is not allowed in FHS 2.3")
+addFilter("W: permissions-file-setuid-bit")
+addFilter("W: permissions-directory-setuid-bit .* is packaged with setuid/setgid bits .*")
+addFilter("W: permissions-missing-verifyscript missing %verify_permissions -e /var/qmail/bin/.*")
```
Why do we need this? Can't we just call the suggested macro?

mbhangui replied:
I need to check this. IIRC, i had some issue using the macro on some version of openSUSE.

mbhangui replied:
This has been done and tested. The noperms variable has been removed to enable permissions.

DerDakon commented on opensuse_buildservice/notqmail-rpmlintrc:
```diff
@@ -0,0 +1,21 @@
+addFilter("W: zero-length /var/qmail/queue/lock/sendmutex")
+setBadness('permissions-unauthorized-file', 0)
+setBadness('permissions-file-setuid-bit', 0)
+setBadness('non-position-independent-executable', 0)
```
Can we just add -fPIC to conf-cc, and maybe -fPIE to conf-load?

mbhangui replied:
Yes for the non-position-independent-executable. Will make that change.

mbhangui replied:
conf-cc and conf-ld are now modified in the spec and rules file

DerDakon commented on opensuse_buildservice/notqmail.dsc:
```diff
@@ -0,0 +1,13 @@
+Debtransform-Files-Tar: debian.tar.gz
+Debtransform-Tar: notqmail-1.08.tar.xz
+Standards-Version: 3.9.1
+Build-Depends: cdbs, debhelper (>= 7.0.50~), gcc, g++, automake, bison (>= 1.2), mime-support, m4, gawk
```
Why do we need g++, automake, bison, m4, and gawk?

mbhangui replied:
Don't think they are needed here as we don't have any calls to awk, yacc, etc in our Makefile. g++, automake also can be removed as I have copy pasted from my project which uses ALT-N's libdkim for DKIM.

DerDakon commented on opensuse_buildservice/notqmail.spec:
```diff
@@ -0,0 +1,423 @@
+
+### SCRIPTLETS ###############################################################################
+%verifyscript
+ID=$(id -u)
```
Why do we need this check?

mbhangui replied:
I need to check this. Will get back on this. In my project I call instcheck to check and fix permissions.

DerDakon reviewed (changes requested):
I have added some comments about things I came accross. Parts of those are simply "I don't know", and many of them "I have not tested", so take it with a grain of salt. You don't need to fix them just to address my comment, if you have just copied things from a template or so then leave them as they are, we can eventually check if these things are actually needed and care for them once we actually know.

mbhangui commented:
The following changes have been made
1. package the systemd unit file notqmail.service in debian builds exactly like the rpm builds
2. run alternatives or update-alternatives command to have notqmail as an alternative mta. This will create link to /var/qmail/bin/sendmail in /usr/lib and /usr/sbin
3. Create /var/qmail/control/me if not present. This allows you to start sending mail righ away after installing notqmail
4. Create alias for root, mailer-daemon and postmaster. This allows you to send mails to root right away after installing notqmail. mails from cron, etc work after this change.
5. On the opensuse build service, debian 7 has been removed

mbhangui commented:
@schmonz There should be a proper email address (something@notqmail.org) in notqmail.dsc, notqmail.spec changelog and debian/changelog. What would be a good address that someone can email in case of issue?

DerDakon commented on opensuse_buildservice/debian/preinst:
```diff
@@ -0,0 +1,66 @@
+
+case "$1" in
+    install|upgrade)
+	rm -f /tmp/notqmail.install /tmp/notqmail.upgrade
```
I find your handling of files in /tmp a bit scary. Even if you remove notqmail.upgrade here I can still create it while the scripts are running to influence what happens later.

Even worse, I can create the logfile as a symlink to something owned by root and can then corrupt it. Does Debian hide all errors of these scripts? Is install-or-upgrade not passed in a sensible way to postinst?

If you really need to communicate these things via file I suggest putting them in /var/qmail, this is something this package controls and that is not writable by the user.

mbhangui replied:
I think in case of upgrade argc is 3 and postinst is called with two arguments where argv1 is configure and argv2 is the version of previous package
in case of installation argc is 2 with argv1 as configure
I will try this. In that case use of files as flags can be eliminated

mbhangui replied:
Even the creation of /tmp/notqmail.log should be removed. It will have the same risk.

mbhangui replied:
done

schmonz commented:
> What would be a good address that someone can email in case of issue?

I've just added a `packages@` alias, forwarding to `dev@`.

mbhangui commented:
We can eliminate instchown in the debian builds if we write a small shim to create the queue. Something like this to create/fix the queue permissions. Here SPLIT and QMAIL comes from conf-split and conf-qmail

```
#!/bin/sh
set -e

create_subdirs()
{
cd $1
count=0
while true
do
        mkdir -p $count
        chmod 700 $count
        count=`expr $count + 1`
        if [ $count -eq SPLIT ] ; then
                break
        fi
done
cd ..
}

create_lockdir()
{
        cd lock
        /bin/rm -f trigger sendmutex
        mkfifo trigger
        chmod 622 trigger
        chown qmails:qmail trigger
        touch sendmutex
        chown qmails:qmail sendmutex
        dd if=/dev/zero of=tcpto count=2 2>/dev/null
        cd ..
}
mkdir -p QMAIL/queue
chmod 750 QMAIL/queue
chown qmailq:qmail QMAIL/queue
cd QMAIL/queue
mkdir -p bounce
mkdir -p info
mkdir -p intd
mkdir -p local
mkdir -p lock
mkdir -p mess
mkdir -p pid
mkdir -p remote
mkdir -p todo

chmod 700 bounce
chmod 700 info
chmod 700 intd
chmod 700 local
chmod 750 lock
chmod 750 mess
chmod 700 pid
chmod 700 remote
chmod 750 todo

create_subdirs info
create_subdirs local
create_lockdir
create_subdirs mess
create_subdirs remote
chown -R qmails:qmail bounce
chown -R qmails:qmail info
chown    qmailq:qmail intd
chown -R qmails:qmail local
chown    qmailq:qmail lock
chown -R qmailq:qmail mess
chown    qmailq:qmail pid
chown -R qmails:qmail remote
chown -R qmailq:qmail todo
```

DerDakon commented:
We have most of the things we need in hier.c and instpackage.c already, but that one is not installed on the target system. I would still leave all the creation to deb and reduce the postinst script to only chown and chmod, which would give us at least correct package ownership tracking, as well as less things to care about in the script. Which means it's just instchown that we care about…

I wonder if we just go and write an "instscript.c" that does basically the same as instchown.c, but instead of executing syscalls outputs the shell commands, and then pipe that into something to be included in the postinst. This way the stuff doesn't clutter the filesystem.

Opinions?

DerDakon commented on opensuse_buildservice/debian/changelog:
```diff
@@ -0,0 +1,35 @@
+notqmail (1.08-1) unstable; urgency=medium
```
Move the "debian" directory up one level, that is where dpkg expects it and can automatically run it from a source tarball.

mbhangui commented:
When it comes to permissions, one needs to have proper permissions for the queue directories and the setuid bit on qmail-queue. debian installs everythhing as root. qmail-queue is easy. Just one chown, chmod and it can be set right. But the queue is a different story. Even for systems other than debian, there is nothing to fix the permissions of the queue unless you have the notqmail source downloaded somewhere. People do screw up things sometimes. Sometime it can be the admin who restores /var/qmail from a backup and finds notqmail not working. 

So there is a need to have something to fix permission problem in rare cases, once in a while to repair the queue or recreate the queue. As we progress with more features tomorrow, there could be a need for a script to do more such things. I have seen one example - qmailctl which does few things

1. stop start qmail
2. build the cdb files for tcpserver
3. queue status
and more

We could decide on one such script/program and have a modular design for it. So that adding functionality should be that we add a plugin, the menu item, options and the actual functionality gets added without hving to modify the base code.

The below is just an example of the qmailctl script

```
Usage: qmailctl {start|stop|condrestart|restart|shut|kill|flush|rotate|reload|stat|queue|pause|cont|cdb|help}
  start -- starts mail service (smtp connection accepted, mail can go out)
   stop -- stops mail service (smtp connections refused, nothing goes out)
restart -- stops and restarts smtp, sends qmail-send a TERM & restarts it
   kill -- Send kill signal to tcpserver, supvervise, qmail-send processes.
   shut -- Shutdown entire svscan service.
  flush -- Schedules queued messages for immediate delivery
 reload -- sends qmail-send HUP, rereading locals and virtualdomains
 status -- status of svscan process & qmail services
 rotate -- rotate all logfiles (sending ALRM to multilog)
  queue -- shows status of queue
  pause -- temporarily stops mail service (connections accepted, nothing goes out)
   cont -- continues paused mail service
    cdb -- rebuild the tcpserver cdb file for smtp, qmtp, qmqp, imap, pop3 and poppass
```

DerDakon commented:
This long exists: http://pyropus.ca/software/queue-repair/ ;)

mbhangui commented:
Just tried the queue repair. It gives syntax error. if I fix that, it gives syntax error in some other part of script. Maybe it needs to be updated for the latest versions of python

```
./queue_repair.py 
  File "/home/mbhangui/Downloads/queue-repair-0.9.0/./queue_repair.py", line 69
    'queue' :           ('qmailq', 'qmail', 0750, 0),
                                               ^
SyntaxError: leading zeros in decimal integer literals are not permitted; use an 0o prefix for octal integers
```

schmonz commented:
In pkgsrc, for similar reasons, I had the qmail package depending on http://www.netmeridian.com/e-huss/queue-fix.tar.gz and running it at post-install time. Maybe that is more suitable here?

mbhangui commented:
> In pkgsrc, for similar reasons, I had the qmail package depending on http://www.netmeridian.com/e-huss/queue-fix.tar.gz and running it at post-install time. Maybe that is more suitable here?

Yes. I use the same. It is written in C. It has saved me many times. The only thing is tht it uses stdio and the coding style will be so different than rest of the code base.

schmonz commented:
Well, maybe it’s not needed. The dependency is long gone from pkgsrc qmail. Sorry I’ve missed this, but what’s the reason it doesn’t work to have the package call `./instchown` at post-install? pkgsrc does it like so: https://github.com/schmonz/pkgsrc-qmail/blob/master/INSTALL

mbhangui commented:
> Well, maybe it’s not needed. The dependency is long gone from pkgsrc qmail. Sorry I’ve missed this, but what’s the reason it doesn’t work to have the package call `./instchown` at post-install? pkgsrc does it like so: https://github.com/schmonz/pkgsrc-qmail/blob/master/INSTALL

Copying instchown has a problem. It bails out if any of the doc or man pages are missing. When we package the man pages, we usually gzip it. Also the docs are moved to a standard directory /usr/share/doc/notqmail instead of /var/qmail/doc.
So on debian, I'm linking /var/qmail/doc to /usr/share/doc/notqmail and /var/qmail/man to /usr/share/man. It stills bails out because the man pages are gzipped. But the code sets the queue permissions before it tries the man pages. So at the moment I just ignore the error by doing
instchown || true

DerDakon commented:
You need to run 2to3 on queue-repair, forgot that.

mbhangui commented:
Yes. It works after doing 2to3

josuah commented:
Improving the tooling of qmail by importing what everybody wanted onto notqmail source... I feel it fits this goal from the wiki:

> Following consensus, merging the best aspects into one implementation

https://github.com/notqmail/notqmail/wiki#what-are-the-projects-goals

DerDakon commented:
When we have the spec files ready, we should add automatic packaging from OBS: https://openbuildservice.org/2021/05/31/scm-integration/

schmonz commented:
Finally got signed up on OBS.

schmonz reviewed (changes requested):
As we're nearing 1.09, probably a good time to rebase this branch and also take advantage of the new `queue-only` argument to `inst{package,chown,check}` as needed.

mbhangui commented:
To use instpackage and instchown, both of these need to be part of the binary package. Till 1.08 we already have instchown in /var/qmail/bin in the binary package. This is being copied to later set the correct permissions of the queue and qmail-queue binary during installation or upgrade. Since there wasn't any way to create queue during the package installation (as source is missing), the queue directories were part of the binary package. This is ok for debian packages. But RPM packages are more sophisticated. You can have non-static directories/files not included as part of the binary package. They can be created later. When you do rpm -V (verification), it will report if any binary, config file has been altered. Having queue as part of the binary package for RPM defeats the purpose. Even for debian, we shouldn't have queue as part of the package. So we need instpackage in /var/notqmail/somewhere. Putting in /var/notmqila/bin doesn't feel right. Though I did it for instchown, I feel both instchown and instpackage needs to be put somewhere? Any thoughts?

schmonz commented:
For pkgsrc I've been [stuffing them](https://github.com/NetBSD/pkgsrc/blob/cc11ec21504359f54163755bd25499654739a12f/mail/qmail/Makefile#L231-L233) into [share/qmail/setup](https://github.com/NetBSD/pkgsrc/blob/cc11ec21504359f54163755bd25499654739a12f/mail/qmail/PLIST#L141-L143) and [calling them from there](https://github.com/NetBSD/pkgsrc/blob/cc11ec21504359f54163755bd25499654739a12f/mail/qmail/INSTALL#L100-L109). Probably there's a location that better meets hier(7) and FHS standards. If you figure out what is, I'll adjust pkgsrc to match. :-)

mbhangui commented:
We can do libexec. `libexec is meant for system daemons and system utilities executed by other programs. That is, the binaries put in this namespaced directory are meant for the consumption of other programs, and are not intended to be executed directly by users.`

schmonz commented:
Do we want our OBS packages to set up `/var/qmail` as a symlink farm, as [an easy and safe step toward FHS](https://github.com/notqmail/notqmail/issues/2#issuecomment-510536117)? FWIW, in July I'll celebrate (gulp) the 20th anniversary of [running qmail in this configuration in production](https://github.com/NetBSD/pkgsrc/commit/f863c03b03bae83242a19a86aa3a723b50a2118a).

mbhangui commented:
Made few changes
1. Added create_rpm, create_debian scripts which can be used by the end user to directly generate RPM or debian packages on the system itself, without using open build service
2. Renamed opensuse_buildservice to open-build-service because the service is now known as open build service
3. A simple guide HOWTO-OBS.md for instructions on how to use obs to generate binary packages for notqmail.
4. updated notqmail.spec, debian/postinst to use instpackage queue-only, instchown. Both RPM and debian builds no longer ship with the queue. The queue is created during first time package installation

mbhangui commented:
> 
> Especially out of place in `libexec` to me is the one that sets the permissions on the `qmail-queue` binary; that seems clearly like an install-time thing, so I would expect that to be the responsibility of whatever is installing that binary (whether it be the package or the `install` target of `Makefile`).
> 

instchown is not required for setting qmail-queue permissions. For RPM, the permission of each and every file is part of the sp,ec file. For debian builds, for setuid programs one has to do chmod 4755 and do dpkg-statoverride thing.  ATM instchown is required to set the queue permissions. instpackage queue-only creates the queue, but with default root permissions. instchown sets the correct ownership. What I really want is a single utility that creates the queue as well as sets the permissions for all subdirs/files/fifo in the queue.

> The one that sets the permissions on the queue seems more reasonable to me, but I'd wish for it to be installed under `bin` or `sbin`, and probably renamed, since it seems more like a "check" or "fix" program (e.g., like NetBSD's [postinstall(8)](https://man.netbsd.org/NetBSD-9.3/postinstall.8)) that an admin can run at any time (e.g., after an upgrade). 

The issue with instpackage is that the only usable option when you don't have the source is the queue-only argument. Using instpackage without that argument and not having the source is a non-working invocation and has no use for the end user. It is for this reason, it doesn't seem right to have this in bin when we have non-working optoins. What I really want is a utility that creates the queue and sets its permission. And such a utility has its use. You may want recreate the queue later. For my own project I use queue-fix. It can create a queue as well as set the right permission and doesn't require the source to be present.

> But now I just saw that Postfix has `libexec/postfix/post-install` [in pkgsrc](https://github.com/NetBSD/pkgsrc/blob/trunk/mail/postfix/PLIST)! I don't know what that does, but if it's not executed by any Postfix program, then it would not be a supporting example of my opinion here. :-)

Linux pretty much uses libexec for such utilities and shared libraries and loadable .so files as in the few examples below

```
/usr/libexec/anaconda:
total 20
-rwxr-xr-x 1 root root  508 Oct 17 21:46 anaconda-pre-log-gen
-rwxr-xr-x 1 root root 1443 Oct 17 21:46 apply-updates
-rwxr-xr-x 1 root root 1481 Oct 17 21:46 log-capture
-rwxr-xr-x 1 root root  301 Oct 17 21:46 run-anaconda
-rwxr-xr-x 1 root root  823 Oct 17 21:46 start-module

/usr/libexec/awk:
total 32
-rwxr-xr-x 1 root root 15880 Jul 19  2023 grcat
-rwxr-xr-x 1 root root 15880 Jul 19  2023 pwcat

/usr/libexec/gcc:
total 0
drwxr-xr-x. 1 root root 4 Dec  5 05:30 x86_64-redhat-linux

/usr/libexec/gcc/x86_64-redhat-linux:
total 0
drwxr-xr-x. 1 root root 132 Dec  9 07:50 13

/usr/libexec/gcc/x86_64-redhat-linux/13:
total 92484
-rwxr-xr-x 1 root root 30692896 Dec  5 05:30 cc1
-rwxr-xr-x 1 root root 32911344 Dec  5 05:30 cc1plus
-rwxr-xr-x 1 root root   347448 Dec  5 05:30 collect2
-rwxr-xr-x 1 root root   215176 Dec  5 05:30 g++-mapper-server
-rwxr-xr-x 1 root root    70536 Dec  5 05:30 liblto_plugin.so
-rwxr-xr-x 1 root root 29594016 Dec  5 05:30 lto1
-rwxr-xr-x 1 root root   857320 Dec  5 05:30 lto-wrapper

```

mbhangui commented:
Another aspect of binary builds I failed to mention is that for every command in /bin, /usr/sbin you should have a man page. Else most package maintainers reject inclusion in the official packages. Debian is pretty much strict about it. One of our aim should be to develop man pages for every binary, script that the user can execute so that notqmail gets accepted by the debian, redhat, archlinux folks. Gentoo too must be having some policy, but my knowlege of gentoo is non-existent to be able to comment on it. For *bsd distributions many of you folks should be able to comment on it as just like gentoo, I don't have much idea.

jlmuir commented:
> What I really want is a utility that creates the queue and sets its permission. And such a utility has its use. You may want recreate the queue later. For my own project I use queue-fix. It can create a queue as well as set the right permission and doesn't require the source to be present.

That sounds good to me.  Could we import your `queue-fix` program?

jlmuir commented:
> Another aspect of binary builds I failed to mention is that for every command in /bin, /usr/sbin you should have a man page.  Else most package maintainers reject inclusion in the official packages. Debian is pretty much strict about it. One of our aim should be to develop man pages for every binary, script that the user can execute so that notqmail gets accepted by the debian, redhat, archlinux folks.

Ah, OK, thanks; good to know.  I could probably write one for your `queue-fix` program if it's agreed that importing it would be the best approach, and the lack of a man page for it is the only reason not to take that approach.

mbhangui commented:
> > What I really want is a utility that creates the queue and sets its permission. And such a utility has its use. You may want recreate the queue later. For my own project I use queue-fix. It can create a queue as well as set the right permission and doesn't require the source to be present.
> 
> That sounds good to me. Could we import your `queue-fix` program?

queue-fix was originally written by Eric Huss. My source is [here](https://github.com/indimail/indimail-mta/blob/master/indimail-mta-x/queue-fix.c). I have adapted it for my project to use substdio and to have a configurable split and bigtodo. I will have to remove -s, -b, -r, -m options to make it usable for notqmail. It can create or fix a corrupted queue. The usage is simple

```
sudo queue-fix /var/qmail/queue
```

If everyone is ok with it, I will have to make a separate PR. But for 1.09 probably we could go ahead without it like we have done for 1.08.

mbhangui commented:
> 
> Ah, OK, thanks; good to know. I could probably write one for your `queue-fix` program if it's agreed that importing it would be the best approach, and the lack of a man page for it is the only reason not to take that approach.

I have already written the [queue-fix man page](https://github.com/indimail/indimail-mta/blob/master/indimail-mta-x/queue-fix.9). But it can be definitely improved as I'm not a native English speaker. So any help in fixing the language, grammar, etc will be helpful. I will remove few things in the man page which will not be relevant to notqmail.

mbhangui commented:
queue-fix along with the man page submitted with PR #267

DerDakon commented:
I force-pushed the branch to clean up the history. All commits that care for the build service files have been preserved. No changes compared to the previous state happened beyond one new commit that was in the master branch now.

schmonz commented on open-build-service/.gitignore:
```diff
@@ -0,0 +1,6 @@
+notqmail.spec
```
Can we add a comment explaining what to keep this synchronized with? (And a matching comment on the other end.)

mbhangui replied:
notqmail.spec is generated from notqmail.spec.in in the Makefile. There are two files that need to be maintained

1. conf-version - This is the notqmail version. e.g. 1.09
2. conf-release - This is the release number for the binary package. e.g. first build could be 1.1

I will update notqmail.spec and other file notqmail.dsc.in, that are generated using rules in the Makefile. Future version could pick this information from the ChangeLog / Release Notes

schmonz commented on open-build-service/HOWTO-OBS.md:
```diff
@@ -0,0 +1,28 @@
+# Open Build Service Steps before release
```
Add a one-line preamble to the effect that this is mostly for the notqmail maintainers, but can also be of use to anyone interested in OBS packaging or building their own.

mbhangui replied:
This has been addressed in the latest commit #[3eaf213](https://github.com/notqmail/notqmail/pull/76/commits/3eaf2138ce40da0831bfc90852faa8b7c2a2c4ff)

schmonz commented on open-build-service/HOWTO-OBS.md:
```diff
@@ -0,0 +1,28 @@
+
+1. Ensure you have following packages installed on the system.
+   These will be build-essential, cdbs, debhelper
+2. Run create\_deban script
```
s/eba/ebia/

mbhangui replied:
This has been addressed in the latest commit #[3eaf213](https://github.com/notqmail/notqmail/pull/76/commits/3eaf2138ce40da0831bfc90852faa8b7c2a2c4ff)

schmonz commented on open-build-service/create_archive:
```diff
@@ -0,0 +1,22 @@
+# $HOME/stage/
+cd ..
+echo "Archiving notqmail-"$version".tar.xz in `pwd`"
+tar \
+	--exclude="notqmail-$version/.git" \
+	--exclude="notqmail-$version/open-build-service"  \
+	--exclude="notqmail-$version/autobuilds"  \
```
Please add a comment pointing to other places that we need to exclude these files (and comments in those places, too). Or, if we can just do it the same way as [our release process](https://notqmail.org/doing-releases/) states, which is currently believed to be the below, that's good enough:

```sh
git archive --prefix=notqmail-${RELEASE}/ -o notqmail-${RELEASE}.tar notqmail-${RELEASE}
```

mbhangui replied:
This script is not needed. Tried git archive and it does the same thing except for inclusion of open-build-service directory (which is not needed for the binary buld)

mbhangui replied:
Will remove this script in the next commit

mbhangui replied:
Instead of removing the script, I have updated it with the git archive command in commit #[3eaf213](https://github.com/notqmail/notqmail/pull/76/commits/3eaf2138ce40da0831bfc90852faa8b7c2a2c4ff)

schmonz commented on open-build-service/create_debian:
```diff
@@ -0,0 +1,21 @@
+#!/bin/sh
```
Add comment and/or rename script to indicate it's not used by OBS directly, but is helpful for testing packaging locally

mbhangui replied:
This has been addressed in the latest commit #[3eaf213](https://github.com/notqmail/notqmail/pull/76/commits/3eaf2138ce40da0831bfc90852faa8b7c2a2c4ff)

schmonz commented on open-build-service/create_rpm:
```diff
@@ -0,0 +1,139 @@
+#!/bin/sh
```
Same as for `create_debian`

schmonz commented on open-build-service/debian/control:
```diff
@@ -0,0 +1,17 @@
+Priority: extra
+Maintainer: notqmail <packages@notqmail.org>
+Standards-Version: 3.9.1
+Homepage: https://github.com/notqmail/notqmail
```
moved to notqmail.org :-) (please search globally in case of other occurrences)

schmonz commented on open-build-service/debian/copyright:
```diff
@@ -0,0 +1,57 @@
+netqmail-1.04
```
Can this be a symlink to top-level `COPYRIGHT` so that it's always in sync? (And so that if we ever misformat the main file, OBS's Debian build will barf and we'll know right away.)

mbhangui replied:
The build fails if this is a symlink. Just tried it on open build service in my local repo

```
[   19s] Scanning //usr/src/packages/SOURCES.DEB/notqmail_1.09.orig.tar.gz...
[   19s] Generating //usr/src/packages/SOURCES.DEB/notqmail_1.09-1.1+5.1.diff
[   19s] Scanning //usr/src/packages/SOURCES/debian.tar.gz...
[   19s] Archive contains a link: debian/copyright -> ../../COPYRIGHT
[   19s] debian transforming failed.
```

schmonz commented on open-build-service/notqmail-permissions.easy:
```diff
@@ -0,0 +1,21 @@
+#
```
A note to the effect that these `notqmail-permissions.*` apply only to OpenSUSE Enterprise rpms, and not other rpms.

mbhangui replied:
This has been addressed by commit #[679054c](https://github.com/notqmail/notqmail/pull/76/commits/679054c6130f8c5ee012714270234bf55cc921d7).

schmonz commented on open-build-service/notqmail-rpmlintrc:
```diff
@@ -0,0 +1,14 @@
+addFilter(".* unversioned-explicit-self-provides")
```
Is this file also OpenSUSE Enterprise only?

mbhangui replied:
No. This is needed in case we provide a virtual library let's say qmail-libs and apt-get install qmail-libs should install libqmail.so.x.y in the system lib directory. We don't provide any shared libraries. This has been removed in the latest commit

schmonz commented on open-build-service/notqmail-rpmlintrc:
```diff
@@ -0,0 +1,14 @@
+addFilter(".* unversioned-explicit-self-provides")
+addFilter(".* non-standard-gid")
+addFilter(".* is not allowed anymore in FHS .*")
```
This reminds us to [do something about FHS sometime](https://github.com/notqmail/notqmail/issues/2#issuecomment-509028214)

mbhangui replied:
Yes. This is about FHS and also make notqmail get accepted by debian and others so that qmailq, qmaill, qmails, qmailr ids and qmail, nofiles gid gets accepted as standard system uid and gids.

schmonz commented on open-build-service/notqmail.dsc.in:
```diff
@@ -0,0 +1,14 @@
+Architecture: any
+Version: @version@-@release@
+Maintainer: notqmail <packages@notqmail.org>
+Homepage: https://github.com/notqmail/notqmail
```
another notqmail.org

mbhangui replied:
This has been addressed in the latest commit #[3eaf213](https://github.com/notqmail/notqmail/pull/76/commits/3eaf2138ce40da0831bfc90852faa8b7c2a2c4ff)

schmonz commented on open-build-service/notqmail.spec.in:
```diff
@@ -0,0 +1,365 @@
+%endif
+
+%description
+notqmail is a community-driven fork of qmail, beginning where netqmail
```
This description is in a few places. Should be updated to the latest from the notqmail.org front page:

> notqmail is the collaborative Open Source successor to [qmail](https://cr.yp.to/qmail.html) and [netqmail](http://netqmail.org/). It begins with stable, compatible, small [releases](https://notqmail.local/releases/) to which existing qmail users can safely [update](https://notqmail.local/install/), and [aims to become](https://notqmail.local/roadmap/) a [more extensible](https://notqmail.local/designs/), [more easily packaged](https://notqmail.local/goals-and-non-goals/), and [more modern](https://notqmail.local/feature-wishlist/) Mail Transport Agent suitable for most needs.

mbhangui replied:
@DerDakon one can use a macro to define a string and then use that macro below %description. Will have to try this out (in the next version) without causing any syntax error

mbhangui replied:
@schmonz , ATM I'm just replacing the description with the updated one as above. For the next version I will look at using sed, head, tail etc to pick the description from README or a common file

mbhangui replied:
This has been addressed in the latest commit #[3eaf213](https://github.com/notqmail/notqmail/pull/76/commits/3eaf2138ce40da0831bfc90852faa8b7c2a2c4ff)

schmonz commented on open-build-service/notqmail.spec.in:
```diff
@@ -0,0 +1,365 @@
+%endif
+rm -f %{buildroot}%{qmaildir}/bin/{qail,elq,pinq,maildirwatch,qsmhook}
+
+%files
```
Probably me being overly anxious, but can we generate these permissions from `hier.c` so they always stay in sync? Maybe a project for later?

schmonz commented on open-build-service/notqmail.spec.in:
```diff
@@ -0,0 +1,365 @@
+fi
+
+# setup notqmail as alternative mta
+alternatives --install /usr/sbin/sendmail mta %{qmaildir}/bin/sendmail 120 \
```
Not a blocker, but for maintenance, will it be possible later to extract common subroutines shared across packages so they can be called uniformly (and without duplication) from .e.g `.deb` and `.rpm` packaging?

mbhangui replied:
This is very much required. I have this in mind. If it is made into a common script, there will be few things that will need to be done depending on whether it is a RPM or a debian system. e.g. updating notqmail as an alternative is different for RPM and debian based system. If we use initscripts in /etc/init.d, /etc/rc*.d, then the alternatives command becomes a bit different on openSUSE systems. So I will have to have few if else conditions in the script. e.g. `if [ -d /etc/debian ] ; then ...`

But, yes making this common makes it less prone to mistakes when changes are made

RPM
```
alternatives --install /usr/sbin/sendmail mta /var/qmail/bin/sendmail 120 \
  --slave /usr/lib/sendmail mta-sendmail /var/qmail/bin/sendmail
```

Debian
```
update-alternatives --install \
  /usr/sbin/sendmail mta /var/qmail/bin/sendmail \
  120 --slave /usr/lib/sendmail mta-sendmail \
  /var/qmail/bin/sendmail
```

schmonz commented on open-build-service/system-users-qmail.conf:
```diff
@@ -0,0 +1,17 @@
+# Type Name ID GECOS [HOME]
```
This is used only by OpenSUSE? Comment plz

mbhangui replied:
Comment has been updated as following. AFAIK, the sysusers_generate_pre macro works only on Fedora and openSUSE > 15 versions. I will test this as part of the next release. If this works for all distributions the useradd commands can be removed from the preinst scripts

```
# Used in notqmail.spec for user creation
# in openSUSE systems
# The sysusers_generate_pre macros generates
# notqmail.pre from system-users-qmail.conf
# This works on openSUSE_LEAP versions >= 15.0 and SLE >= 15
# %if 0%{?suse_version} >= 1500 || 0%{?sles_version} >= 15
```

schmonz reviewed (changes requested):
At a glance, without trying to deeply understand these packaging systems, it is plainly evident how much effort you put into understanding these packaging systems. Thank you for going to the trouble: it will help our users (and especially future users) quite a bit. We see some opportunities to lower the future maintenance cost of this packaging code but mostly no urgency, we can do them at leisure as we continue to live with this integrated packaging automation.

DerDakon commented on open-build-service/notqmail.spec.in:
```diff
@@ -0,0 +1,365 @@
+  %run_permissions
+%endif
+%endif
+/var/qmail/bin/instpackage queue-only
+/var/qmail/bin/instchown queue-only
```
That will break as they are in libexec, no?

mbhangui replied:
good catch - fixed. Will get updated in the next commit.

mbhangui replied:
This has been addressed in the latest commit #[3eaf213](https://github.com/notqmail/notqmail/pull/76/commits/3eaf2138ce40da0831bfc90852faa8b7c2a2c4ff)

mbhangui commented on open-build-service/notqmail.spec.in:
```diff
@@ -0,0 +1,365 @@
+  %{__mkdir_p} %{buildroot}%{_unitdir}
+  install -m 644 qmail-send.service %{buildroot}%{_unitdir}/%{name}.service
+%endif
+rm -f %{buildroot}%{qmaildir}/bin/{qail,elq,pinq,maildirwatch,qsmhook}
```
Done - will be part of the next comment

mbhangui commented on open-build-service/notqmail-rpmlintrc:
```diff
@@ -0,0 +1,14 @@
+addFilter(".* unversioned-explicit-self-provides")
+addFilter(".* non-standard-gid")
+addFilter(".* is not allowed anymore in FHS .*")
+addFilter(".* invalid-license .*")
```
These macros supress rpmlint errors. On SLES, it is impossibel to build without supressing these errors. Lot of these are noise because qmail does things different than standard linux distributions. Let me explain each of those lines

# addFilter(".* unversioned-explicit-self-provides")
The unversioned-explicit-self provides can be removed. That error happens when we provide a library and say
```
Provides: qmail-libs
```
RPM and debian generates automatically any shared library provided by package. But sometimes you want your own name to be provided. e.g. MySQL provides mysql-libs but the actual library name provided is libmysqlclient.so.x.x

# addFilter(".* non-standard-gid")

When we install packages the binaries go into /var/qmail/bin with the group as `qmail`. qmail-queue gets installed with owner as `qmailq:qmail`. At the moment none of the current linux distributions will recognize these uids to be valid for binaries outside the user's home directories

# addFilter(".* is not allowed anymore in FHS .*")

This is absolutely needed on the open build service because putting binaries in /var/qmail/bin goes against FHS. This error is printed for each and every binary put in /var/qmail/bin

# addFilter(".* invalid-license .*")

I'm not sure from which file the open build service rpmlint program picks up the list of valid licenses. In our case we have the following licence which the build service complains

```
License: CC-PDDC
```

Note: I have removed `addFilter(".* unversioned-explicit-self-provides")` from the rpmlintrc file and will be part of the next commit

mbhangui commented on open-build-service/notqmail.spec.in:
```diff
@@ -0,0 +1,365 @@
+
+%package doc
+Summary: Documentations for the %{name} package
+%if %{undefined suse_version} && %{undefined sles_version}
+Group: System Environment/Base
+%else
+Group: Productivity/Networking/Email/Servers
+%endif
```
Can be done. We can have all definitions at the top and after all definitions we can avoid having these %if %else branches. That will make the spec file more readable. I will put this in my action items for next release (after 1.09)

mbhangui replied:
I will look into this for the next release after 1.09

mbhangui commented on open-build-service/notqmail.spec.in:
```diff
@@ -0,0 +1,365 @@
+%if 0%{?suse_version} >= 1500 || 0%{?sles_version} >= 15
+  %sysusers_generate_pre %{S:4} notqmail
+%endif
+conf_cc=`head -1 conf-cc`
```
Done - Modified. Will be part of next commit

mbhangui commented on open-build-service/notqmail.spec.in:
```diff
@@ -0,0 +1,365 @@
+mv %{buildroot}%{qmaildir}/man/man? %{buildroot}%{_mandir}
+mv %{buildroot}%{qmaildir}/doc/* %{buildroot}%{nq_docdir}
+%{__cp} COPYRIGHT %{buildroot}%{nq_docdir}
+rmdir --ignore-fail-on-non-empty %{buildroot}%{qmaildir}/doc
+rmdir --ignore-fail-on-non-empty %{buildroot}%{qmaildir}/man
```
Done - Modified. Will be part of next commit

mbhangui replied:
This has been addressed in the latest commit #[3eaf213](https://github.com/notqmail/notqmail/pull/76/commits/3eaf2138ce40da0831bfc90852faa8b7c2a2c4ff)

mbhangui commented on open-build-service/notqmail.spec.in:
```diff
@@ -0,0 +1,365 @@
+%if 0%{?suse_version} >= 1120
+  %{__mkdir_p} %{buildroot}%{_sysconfdir}/permissions.d/
+  install -m 644 %{S:1} %{buildroot}%{_sysconfdir}/permissions.d/%{name}-permissions
+  install -m 644 %{S:2} %{buildroot}%{_sysconfdir}/permissions.d/%{name}-permissions.secure
```
In reality none of the files are actually needed except for the SLES systems. But without qmail-queue being setuid, we cannot queue any mail without making insecure permissions for the queue directory. I have it there because generally I see distributions including all three (easy, secure and paranoid). These files are bit like the debian dpkg-override. I'm in two minds about removing the paranoid file. Having the file reminds me that there are three security settings for permissions. But yes, it is uselessly sitting out there.

mbhangui replied:
This has been addressed as part of commit #[679054c](https://github.com/notqmail/notqmail/pull/76/commits/679054c6130f8c5ee012714270234bf55cc921d7)

mbhangui commented on open-build-service/notqmail.spec.in:
```diff
@@ -0,0 +1,365 @@
+%endif
+%if "%{?_unitdir}" != ""
+  make -s qmail-send.service
+  %{__mkdir_p} %{buildroot}%{_unitdir}
+  install -m 644 qmail-send.service %{buildroot}%{_unitdir}/%{name}.service
```
Done - will be part of the next commit

mbhangui replied:
This has been addressed in the latest commit #[3eaf213](https://github.com/notqmail/notqmail/pull/76/commits/3eaf2138ce40da0831bfc90852faa8b7c2a2c4ff)
